### PR TITLE
Allow nesting of OrderMutex

### PR DESCRIPTION
**Pull request by philbirt**
Currently, if you have a method that acquires an OrderMutex lock, which calls another method that attempts to lock the order, you will get a LockFailed error.

It would be ideal if we could lock the order, and if further down the chain we attempt to do it again, it would know that we have already acquired it and to just yield to the block without erroring.

This is my attempt at making it work. The spec is a little hairy, so if anyone has tips on testing multi-threaded stuff, let me know!

### CODE REVIEW

**Comment by magnusvk** [comment]
This looks reasonable to me. Would love to get @jordan-brough's eyes on it. The spec failures looked unrelated.

**Comment by jhawthorn** [comment]
I'm not sure this is a good approach to begin with.

In addition, the new test is broken. It doesn't testing for a yeild (b variable is unused).

> The spec failures looked unrelated.

Why would you think this? The failing test is called "Spree::Api::BaseController lock_order with an existing lock returns a 409 conflict" and has failed three times in a row.

**Comment by jordan-brough** [comment]
I think this approach looks great.  It does seem like we ought to be able to make those tests a little nicer.

> I'm not sure this is a good approach to begin with

I actually really like OrderMutex (though I'm biased since I created it :) ).  It's worked really well for us and has been exceedingly pragmatic.  I'm definitely interested in better solutions, and if there are ones that are feasible for Solidus right now that would be great to hear about.  There's the whole 'use database transactions, duh' solution that we obviously ought to be doing more of, but trying to use that as a system-wide locking scheme in Solidus is just not feasible right now as far as I've been able to tell.

In other words:  Please open a github issue. ;)

File: core/app/models/spree/order_mutex.rb [hunk 1]
```diff
@@ -14,20 +14,25 @@ class << self
...
-        # limit the maximum lock time just in case a lock is somehow left in place accidentally
-        expired.where(order: order).delete_all
-
-        begin
-          order_mutex = create!(order: order)
-        rescue ActiveRecord::RecordNotUnique
-          error = LockFailed.new("Could not obtain lock on order #{order.id}")
-          logger.error error.inspect
-          raise error
+        if @orderMutexThreadId ==  Thread.current.object_id
```
Comment (jordan-brough) [inline core/app/models/spree/order_mutex.rb]: nitpick: extra space here

**Comment by jordan-brough** [comment]
@jhawthorn I'm assuming when you say "I'm not sure this is a good approach to begin with" that the "this" is OrderMutex.  If the "this" is "nested OrderMutexes" then please ignore my whole previous comment.  I'm not completely sure whether we should allow nesting, but I can't think of a solid reason why not at the moment.

Thinking about this a bit more, I wonder if we need to require that a transaction not be open when the order mutex is created.  If a transaction were already open then any other attempts to acquire the mutex would block, where so far we've been trying to get them to fail immediately.

**Comment by jhawthorn** [comment]
@jordan-brough No. This code here in this PR is not a good approach. I also don't think OrderMutex is a good approach, but I wasn't bringing that up here.

This still isn't passing tests, so I don't know how you are declaring this approach "great".

File: core/app/models/spree/order_mutex.rb [hunk 2]
```diff
@@ -14,20 +14,25 @@ class << self
...
+            raise error
+          end
+
+          yield
         end
-
-        yield
-
       ensure
+        @orderMutexThreadId = nil
```
Comment (jordan-brough) [inline core/app/models/spree/order_mutex.rb]: I don't think this will work correctly, will it?  If another Thread comes into this and exits then it'll clear this for all threads.

Comment (jordan-brough) [inline core/app/models/spree/order_mutex.rb]: Perhaps we just need to add `if @orderMutexThreadId == Thread.current.object_id` to this?

Comment (jordan-brough) [inline core/app/models/spree/order_mutex.rb]: Oh, also -- the first nested call in the same thread will clear this.  We probably need to move it up to the prior `begin-rescue`.

Comment (jordan-brough) [inline core/app/models/spree/order_mutex.rb]: (And make sure we have specs for both of those cases)

File: core/spec/models/spree/order_mutex_spec.rb [hunk 3]
```diff
@@ -38,6 +36,23 @@
     end
   end
 
+  context "when another thread requests lock on already locked order" do
+    it "raises a LockFailed error" do
+      Thread.new do
+        Spree::OrderMutex.with_lock!(order) {|o| sleep(3) }
```
Comment (jordan-brough) [inline core/spec/models/spree/order_mutex_spec.rb]: @philbirt I know you already called out that the specs could use some improvement, but we'll find a better way to do this other than sleeping right?  I'm no threading expert either but happy to pair up and brainstorm ideas if that's helpful. Just lmk.

**Comment by jordan-brough** [comment]
> I also don't think OrderMutex is a good approach, but I wasn't bringing that up here.

OK, cool. Sorry for my misunderstanding there.

> This code here in this PR is not a good approach

It may have some implementation flaws but the general approach seems good to me.  Allowing nested mutexes makes me slightly nervous but I can't put my finger on why still.  Is there anything specific that's making you dislike the approach?

> I think it's early to thumbs up this approach

Just to be clear -- no one has given it a thumbs-up yet.

**Comment by jordan-brough** [comment]
One thing to think about regarding the current approach -- if you've already got a lock then any threads you spawn won't be able to acquire a lock. That actually seems correct to me because the whole point of the OrderMutex is to prevent concurrent modification of an order.  And I think that's just a limitation of where we're at with concurrency in Solidus, but it might not be intuitive (it wasn't to me at first).  Thoughts on that?  If nothing else it's worth calling out in some comments.

Btw I think that sort of complexity is part of what makes me nervous about allowing nesting. Maybe it's fine but are there other options we should explore?  @philbirt is there any reasonable way to satisfy #502 without nesting?

**Comment by philbirt** [comment]
@jordan-brough Yeah there are definitely ways to satisfy https://github.com/solidusio/solidus/pull/502 without doing this at all. First and easiest, I could just leave the refactoring to use common code alone and accomplish that later. I'm sure there are other alternatives, but that seems the most sane to me.

@jhawthorn I just want to be clear that I didn't come into this thinking this was the only or best approach. Hence why I said "this is my attempt at making it work". I also didn't really need to do it -- I thought and still think it's reasonable-ish, and I thought it was a desirable feature, so I attempted to make it work.

I agree some seemingly unrelated ones are still failing and I am looking into it, because they most definitely are related. I, like you, have zero desire to have a red master or do anything risky with this, as we are heavily reliant on it locking when we expect it to and releasing the lock when we expect it to. This isn't a deal breaker for us to get in quickly (or at all for that matter), so let's take our time and get it right. I would just really appreciate some feedback on how we could make it better and specifically what is "not good" about it.

**Comment by magnusvk** [comment]
@philbirt @jhawthorn Any more thoughts on this? It looks like #502 doesn't require this as it currently stands. Should we close this or is this still something we should go after? IIRC @jordan-brough had code for this that worked that we could try and bring in here. But I wouldn't be opposed to just closing this on account of "we are not sure enough we'll get it right."

**Comment by jordan-brough** [comment]
Here's the draft I whipped up and showed to @jhawthorn and @cbrunsdon: https://gist.github.com/jordan-brough/a0b9f208b6b5b445dfc4  We took a look together and believe that it should work correctly (without having written any specs and w/ just 15min of looking at it).

However, I believe we ended up not needing the nesting after all, and it would be nice to avoid that complexity until we find we really do need it.  So I'm going to close this PR for now and someone can re-open if they think we should pursue this further.
